concrete-core/src/backends/core/private/crypto/ggsw/fourier.rs

```diff
@@ -0,0 +1,725 @@
+    ///     Complex64::new(0., 0.),
+    ///     PolynomialSize(9),
+    ///     GlweSize(7),
+    ///     DecompositionLevelCount(3),
+    ///     DecompositionBaseLog(4),
+    /// );
+    /// ggsw.par_level_matrix_iter_mut()
+    ///     .for_each(|mut level_matrix| {
+    ///         for mut rlwe in level_matrix.row_iter_mut() {
+    ///             rlwe.as_mut_tensor()
+    ///                 .fill_with_element(Complex64::new(0., 0.));
+    ///         }
+    ///     });
+    /// assert!(ggsw
+    ///     .as_tensor()
+    ///     .iter()
+    ///     .all(|a| *a == Complex64::new(0., 0.)));
+    /// assert_eq!(ggsw.level_matrix_iter_mut().count(), 3);
+    /// ```
+    #[cfg(feature = "multithread")]
+    pub fn par_level_matrix_iter_mut(
+        &mut self,
+    ) -> impl IndexedParallelIterator<Item = GgswLevelMatrix<&mut [<Self as AsRefTensor>::Element]>>
+    where
+        Self: AsMutTensor,
+        <Self as AsMutTensor>::Element: Sync + Send,
+    {
+        let chunks_size = self.poly_size.0 * self.rlwe_size.0 * self.rlwe_size.0;
+        let poly_size = self.poly_size;
+        let rlwe_size = self.rlwe_size;
+        self.as_mut_tensor()
+            .par_subtensor_iter_mut(chunks_size)
+            .enumerate()
+            .map(move |(index, tensor)| {
+                GgswLevelMatrix::from_container(
+                    tensor.into_container(),
+                    poly_size,
+                    rlwe_size,
+                    DecompositionLevel(index + 1),
+                )
+            })
+    }
+
+    /// Fills a GGSW ciphertext with the fourier transform of a GGSW ciphertext in
+    /// coefficient domain.
+    pub fn fill_with_forward_fourier<InputCont>(
+        &mut self,
+        coef_ggsw: &StandardGgswCiphertext<InputCont>,
+        buffers: &mut FourierBuffers<Scalar>,
+    ) where
+        Cont: AsMutSlice<Element = Complex64>,
+        StandardGgswCiphertext<InputCont>: AsRefTensor<Element = Scalar>,
+        Scalar: UnsignedTorus,
+    {
+        // We retrieve a buffer for the fft.
+        let fft_buffer = &mut buffers.fft_buffers.first_buffer;
+        let fft = &mut buffers.fft_buffers.fft;
+
+        // We move every polynomials to the fourier domain.
+        let iterator = self
+            .tensor
+            .subtensor_iter_mut(self.poly_size.0)
+            .map(|t| FourierPolynomial::from_container(t.into_container()))
+            .zip(
+                coef_ggsw
+                    .as_tensor()
+                    .subtensor_iter(coef_ggsw.polynomial_size().0)
+                    .map(|t| Polynomial::from_container(t.into_container())),
+            );
+        for (mut fourier_poly, coef_poly) in iterator {
+            fft.forward_as_torus(fft_buffer, &coef_poly);
+            fourier_poly
+                .as_mut_tensor()
+                .fill_with_one((fft_buffer).as_tensor(), |a| *a);
+        }
+    }
+
+    pub fn external_product<C1, C2>(
```

Comment:
maybe we could add a little doc for that function like there are above?

Reply:
Yes, I should copy the doc from the other external_product, I wanted to complete the implementation befroe doing so, then forgot.

Reply:
My take on it would be: if we add something, it must be to help developers understand something which is not obvious. This can be the case for general architecture of a module, or technical details which are difficult to understand. Otherwise, I would try to reduce the amount of documentation in `private`. The doc which is available is a relic of the former public API, I would not say it is wise to keep the same level of doc in the private level. It will take time to maintain, and will probably get outdated very soon.

Reply:
I won't add antything then, it is assumed that someone mingling with private knows what an external product is at a high level, and the code in itself is described in the following comments.